12-offer-encoding.md

```diff
@@ -0,0 +1,1053 @@
+   * [`u16`:`len`]
+   * [`len*byte`:`address`]
+
+## Requirements
+
+A writer of an invoice:
+  - MUST set `created_at` to the number of seconds since Midnight 1
+    January 1970, UTC when the offer was created.
+  - MUST set `payment_hash` to the SHA2 256-bit hash of the
+    `payment_preimage` that will be given in return for payment.
+  - MUST set (or not set) `send_invoice` the same as the offer.
+  - MUST set `offer_id` to the id of the offer.
+  - MUST specify exactly one signature TLV: `signature`.
+    - MUST set `sig` to the signature using `node_id` as described in [Signature Calculation](#signature-calculation).
+  - if the chain for the invoice is not solely bitcoin:
+    - MUST specify `chains` the invoice is valid for.
+  - otherwise:
+    - the bitcoin chain is implied as the first and only entry.
+  - if it has bolt11 features:
+    - MUST set `features` to the bitmap of features.
+  - if the invoice corresponds to an offer with `recurrence`:
+    - MUST set `recurrence_basetime` to the start of period #0 as calculated
+      by [Period Calculation](#offer-period-calculation).
+    - if it sets `relative_expiry`:
+      - MUST NOT set `relative_expiry` `seconds_from_creation` more than the number of seconds after `created_at` that payment for this period will be accepted.
+  - otherwise:
+    - MUST not set `recurrence_basetime`.
+  - if the expiry for accepting payment is not 7200 seconds after `created_at`:
+    - MUST set `relative_expiry` `seconds_from_creation` to the number of
+      seconds after `created_at` that payment of this invoice should not be attempted.
+  - if the `min_final_cltv_expiry` for the last HTLC in the route is not 18:
+    - MUST set `min_final_cltv_expiry`.
+  - if it accepts onchain payments:
+    - MAY specify `fallbacks`
+    - MUST specify `fallbacks` in order of most-preferred to least-preferred
+      if it has a preference.
+    - for the bitcoin chain, it MUST set each `fallback_address` with
+      `version` as a valid witness version and `address` as a valid witness
+      program
+  - if it is connected only by private channels:
+    - MUST include a `blinded_path` containing one or more paths to the node.
+  - otherwise:
+    - MAY include `blinded_path`.
+  - if it includes `blinded_path`:
+    - MUST specify `path` in order of most-preferred to least-preferred if
+      it has a preference.
+    - MUST include `blindedpay` with exactly one `payinfo` for
+      each `onionmsg_path` in `blinded_path`, in order.
+    - if it includes `blinded_capacities`:
+      - MUST include exactly one `incoming_msat` (in millisatoshis) per `path`, reflecting the expected minimum amount that can be sent through the path.
+    - SHOULD ignore any payment which does not use one of the paths.
+  - otherwise:
+    - MUST NOT include `blinded_payinfo`.
+  - MUST specify `amount`.`msat` in multiples of the minimum lightning-payable unit
+    (e.g. milli-satoshis for bitcoin) for the first `chains` entry.
```

Comment:
I'm probably missing something, but wouldn't the invoice be kind of pointless for other chains given that it only has the amount for one chain?
For the invoice request and offer messages, I can see how having multiple chains make sense, but for the individual invoice, I fail to see how more than one chain makes sense.

Reply:
I agree; by the time we issue an invoice, let's keep it to a single chain.  Will fix

Reply:
There can be multiple chains which use the same currency, though I expect it will be a single value in practice...

Reply:
I guess for example you could have btc and lbtc, for example. But I think that's too specific of a use case and should be handled at the application level in OSI terms. 
At the protocol level they're different assets  and their msats/amouts cannot be directly compared.
In the future, I hope, lightning will also support multi-asset chains, so I'm strongly against the notion of two chains being considered the same asset at the protocol level. In my opinion each user should be free to elect which assets are equivalent (aka 1:1 price) or whatever price they want in their offers.

Reply:
I still think invoices should be specific to a chain and nobody will use it.
Let's say both recipient accept either BTC or LBTC and the recipient accepts BTC and LBTC as being 1:1 price perfectly equivalent assets at all times (I think he shouldn't, but I cannot think of another use case).  Still in this case, if payer is going to chose, he can decide in the previous step, can't he?
There's no advantage that I can see in the user choosing the chain later.
I guess it's not terrible to allow it, I just don't see the point.
Perhaps some example can help me understand.

Reply:
The harm is in the added complexity. Why use a list if we know that it will always be of size 1?